README.md

# Cybersecurity News Website

This repository contains the code for a cybersecurity news website. The website aggregates articles from various cybersecurity news sources, assigns scores based on the source's credibility, freshness, and engagement, and displays the articles in order of their overall score. The website also uses a machine learning model to predict the source of an article based on its summary.

## Code Overview

### Flask Application (Backend)

The Flask application provides the backend for the website. It sets up the routes for the website, retrieves articles from different sources, calculates various scores for each article, and sorts the articles by the overall score. It also provides a route to clear the cache.

### Feed Parser (Data Collection)

The script parses RSS feeds from various cybersecurity news sources, assigns scores based on the source's credibility, freshness, and engagement, and stores the articles in a MongoDB database.

### Data Cleaner (Data Preprocessing)

The script reads the articles from the database, removes duplicates, cleans the text of the articles, and stores the cleaned articles back in the database.

### Model Trainer (Machine Learning)

The script preprocesses the data, trains a Logistic Regression model with hyperparameter tuning and class balancing to predict the source of an article based on its summary, validates the model, and saves the model and necessary transformers for later use.

## Setup Instructions

The `setup.py` file is used to automate the installation of the necessary packages and dependencies to run the cybersecurity news aggregator. 

Here's how you can use it:

1. Download or clone the repository to your local machine.

2. Open a terminal window and navigate to the directory where the `setup.py` file is located.

3. Run the following command:
   
   ```python
   pip install .
   ```

1. Clone the repository.
2. Install the necessary dependencies.
3. Set up a MongoDB database.
4. Run the feed parser to collect data.
5. Run the data cleaner to preprocess the data.
6. Run the model trainer to train and validate the machine learning model.
7. Run the Flask application to start the website.
8. You can automate this process by running "python news" in the command line


reader/news
import subprocess

# Run reader.py
subprocess.run(['python', 'reader.py'])

# Run datacleaner.py
subprocess.run(['python', 'data_cleaner.py'])

# Run model.py
subprocess.run(['python', 'model.py'])

# Run app.py
subprocess.run(['python', 'app.py'])


reader/model.py
import pickle
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import GridSearchCV
from data_cleaner import clean_data
from data_cleaner import clean_text, preprocess_text

def train_and_validate_model():
    # Call the cleaning function from the reader.py file
    cleaned_data = clean_data()
    
    # Convert text data to numerical data using TF-IDF
    vectorizer = TfidfVectorizer()
    X = vectorizer.fit_transform(cleaned_data['summary'])
    y = cleaned_data['source']
    
    # Save the vectorizer for later use
    with open('vectorizer.pkl', 'wb') as f:
        pickle.dump(vectorizer, f)
        
    # Convert categorical labels to numerical values
    label_encoder = LabelEncoder()
    y = label_encoder.fit_transform(y)
    
    # Save the label encoder for later use
    with open('label_encoder.pkl', 'wb') as f:
        pickle.dump(label_encoder, f)
    
    # Split the data into training and validation sets
    X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)
    
   # Train the model with hyperparameter tuning and class balancing
    parameters = {'C': [0.1, 1, 10], 'penalty': ['l1', 'l2']}
    model = GridSearchCV(LogisticRegression(class_weight='balanced', solver='liblinear', max_iter=1000), parameters, cv=5)
    model.fit(X_train, y_train)

    
    # Save the model for later use
    with open('model.pkl', 'wb') as f:
        pickle.dump(model, f)
    
    # Validate the model
    y_pred = model.predict(X_val)
    
    # Print the evaluation metrics
    print(f'Accuracy: {accuracy_score(y_val, y_pred)}')
    print(f'Precision: {precision_score(y_val, y_pred, average="weighted", zero_division=0)}')
    print(f'Recall: {recall_score(y_val, y_pred, average="weighted")}')
    print(f'F1 Score: {f1_score(y_val, y_pred, average="weighted")}')

def load_model(model_file):
    with open(model_file, 'rb') as f:
        model = pickle.load(f)
    return model

model = load_model('model.pkl')

def preprocess_article(article):
    # Apply the same preprocessing steps as used during training
    cleaned_text = clean_text(article)  # Use the cleaning function from the data_cleaner.py file
    preprocessed_text = preprocess_text(cleaned_text)  # Apply additional preprocessing steps if needed
    return preprocessed_text

def predict_category(article):
    preprocessed_article = preprocess_article(article)
    prediction = model.predict([preprocessed_article])[0]
    return prediction

if __name__ == '__main__':
    train_and_validate_model()


reader/reader.py
import feedparser
from urllib.parse import urlparse
from pymongo import MongoClient
from datetime import datetime
from dateutil import parser
import pytz
import requests
from bs4 import BeautifulSoup
from data_cleaner import clean_text
import random


def get_credibility_score(source):
    credibility_scores = {
        'The Hacker News': 10,
        'Graham Cluley': 9,
        'Naked Security': 8.5,
        'The Last Watchdog': 8,
        'The State of Security - Tripwire': 7.5,
        'CSO Online': 7,
        'Infosecurity Magazine': 6.5,
        'Cyber Magazine': 6,
        'Security Weekly': 5.5,
        'IT Security Guru': 5
    }
    return credibility_scores.get(source, 5)

def get_freshness_score(published_time):
    current_time = datetime.now(pytz.utc)
    published_time = datetime(*published_time[:6], tzinfo=pytz.utc)
    time_difference = current_time - published_time
    hours_difference = time_difference.total_seconds() / 3600
    return max(10 - hours_difference / 24, 0)

def get_engagement_score():
    return 5

def get_overall_score(credibility, freshness, engagement):
    return credibility * 0.5 + freshness * 0.3 + engagement * 0.2

def extract_thumbnail(entry):
    thumbnail = None

    # Check if the entry has a 'media_thumbnail' field
    if 'media_thumbnail' in entry:
        for media in entry['media_thumbnail']:
            if 'url' in media and media['url']:
                thumbnail = media['url']
                break

    # Check if the entry has 'enclosures'
    elif 'enclosures' in entry:
        for enclosure in entry['enclosures']:
            if 'type' in enclosure and enclosure['type'].startswith('image'):
                if 'href' in enclosure and enclosure['href']:
                    thumbnail = enclosure['href']
                    break

    # Check if the entry has 'content' with 'value' field
    elif 'content' in entry:
        for content in entry['content']:
            if (
                'type' in content and content['type'].startswith('image')
                and 'value' in content and content['value']
            ):
                soup = BeautifulSoup(content['value'], 'html.parser')
                img_tags = soup.find_all('img')
                if img_tags:
                    thumbnail = img_tags[0].get('src')
                    break

    # Retrieve a random image from Pixabay
    if not thumbnail:
        api_key = "your_api_key"
        query = ["technology", "computing", "cybersecurity", "encryption", "firewall", "malware", "phishing", 
                 "ransomware", "spyware", "threat", "vulnerability", "intrusion", "forensics", "network", 
                 "data", "privacy", "protection", "breach", "alert", "defense", "identity", "risk", "scam", 
                 "security", "attack", "cybercrime", "hacker", 
                 "safe", "secure", "system", "information", "internet", "virus", "web", "password"]
        thumbnail = get_random_image_from_pixabay(api_key, query)

    return thumbnail

# Create a set to store used image URLs
used_images = set()

def get_random_image_from_pixabay(api_key, query):
    url = "https://pixabay.com/api/"
    params = {
        "key": api_key,
        "q": query,
        "orientation": "horizontal",
        "per_page": 20
    }
    response = requests.get(url, params=params)
    if response.status_code == 200:
        result = response.json()
        if "hits" in result and len(result["hits"]) > 0:
            # Select a random image and check if it's been used
            image = random.choice(result["hits"])
            while image["largeImageURL"] in used_images and len(used_images) < len(result["hits"]):
                image = random.choice(result["hits"])
            used_images.add(image["largeImageURL"])
            return image["largeImageURL"]
    return None


# List of queries
query = [
    "technology"
          ]

# Usage example
api_key = "38297032-8bacc7c47db30f0e9b5ac6218"

for query in query:
    image_url = get_random_image_from_pixabay(api_key, query)
    if image_url:
        print(f"Random Image URL for query '{query}': {image_url}")
    else:
        print(f"No image found for query '{query}'.")


def parse_feeds(sources, db):
    all_articles = []
    for source_name, source_url in sources.items():
        print(f'Parsing feed from {source_name}...')

        feed = feedparser.parse(source_url)

        for entry in feed.entries:
            print('Title: ', entry.title)
            print('Link: ', entry.link)
            print('Published: ', entry.published)
            print('Summary: ', entry.summary)
            print()

            credibility = get_credibility_score(source_name)
            freshness = get_freshness_score(entry.published_parsed)
            engagement = get_engagement_score()
            overall_score = get_overall_score(credibility, freshness, engagement)

            # Extract the title and summary using the provided entry.get() method
            original_title = entry.get('title', '')
            summary = entry.get('summary', '')

            # Clean the title using the clean_text() function from data_cleaner.py
            cleaned_title = clean_text(original_title)

            # Parse the summary using BeautifulSoup to extract the text content
            soup = BeautifulSoup(summary, 'html.parser')
            cleaned_summary = soup.get_text()

            # Extract the thumbnail image URL
            thumbnail = extract_thumbnail(entry)

            # Parse the published date using dateutil.parser
            published_date = parser.parse(entry.published)
            published_iso = published_date.isoformat()

            article = {
                'source': source_name,
                'title': original_title,  # Use the original title here
                'cleaned_title': cleaned_title,  # You can still store the cleaned title if you need it
                'link': entry.link,
                'published': published_iso,
                'summary': cleaned_summary,
                'credibility_score': credibility,
                'freshness_score': freshness,
                'engagement_score': engagement,
                'overall_score': overall_score,
                'thumbnail': thumbnail
            }
            all_articles.append(article)

    # Sort all_articles by 'published' field
    all_articles.sort(key=lambda article: article['published'], reverse=True)

    # Insert sorted articles into database
    db.articles.insert_many(all_articles)

def run_parser():
    client = MongoClient('localhost', 27017)
    db = client['cybersecurity_news']

    sources = {
        'The Hacker News': 'https://thehackernews.com/feeds/posts/default',
        'Graham Cluley': 'https://feeds.feedburner.com/GrahamCluleysBlog',
        'Naked Security': 'https://nakedsecurity.sophos.com/feed/',
        'The Last Watchdog': 'https://www.lastwatchdog.com/feed/',
        'The State of Security - Tripwire': 'https://www.tripwire.com/state-of-security/feed/',
        'CSO Online': 'https://www.csoonline.com/index.rss',
        'Infosecurity Magazine': 'https://www.infosecurity-magazine.com/rss/news/',
        'Cyber Magazine': 'https://cyber-magazine.nl/feed/',
        'Security Weekly': 'https://securityweekly.com/feed/',
        'IT Security Guru': 'https://www.itsecurityguru.org/feed/'
    }

    parse_feeds(sources, db)  # Pass the 'db' argument

if __name__ == '__main__':
    run_parser()

reader/data_cleaner.py
import pandas as pd
import re
import nltk
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from pymongo import MongoClient


# Download nltk data if not already downloaded
nltk.download('stopwords', quiet=True)
nltk.download('wordnet', quiet=True)


def clean_text(text):
    # Convert text to lowercase
    text = text.lower()

    # Remove punctuation
    text = re.sub(r'[^\w\s]', '', text)

    # Tokenize text
    words = text.split()

    # Remove stopwords and lemmatize words
    lemmatizer = WordNetLemmatizer()
    words = [lemmatizer.lemmatize(word) for word in words if word not in set(stopwords.words('english'))]

    # Return cleaned text
    return ' '.join(words)


def preprocess_text(text):
    # Convert text to lowercase
    text = text.lower()

    # Remove punctuation
    text = re.sub(r'[^\w\s]', '', text)

    # Tokenize text
    words = text.split()

    # Remove stopwords and lemmatize words
    lemmatizer = WordNetLemmatizer()
    words = [lemmatizer.lemmatize(word) for word in words if word not in set(stopwords.words('english'))]

    # Return preprocessed text
    return ' '.join(words)


def clean_data():
    client = MongoClient('localhost', 27017)
    db = client['cybersecurity_news']

    data = pd.DataFrame(list(db.articles.find()))

    data.drop_duplicates(subset=['title', 'link'], keep='first', inplace=True)
    data.dropna(subset=['summary'], inplace=True)  # Drop documents with empty summaries

    # Remove stop word-only documents
    stop_words = set(stopwords.words('english'))
    data['summary_processed'] = data['summary'].apply(lambda x: preprocess_text(x))
    data['word_count'] = data['summary_processed'].apply(lambda x: len([word for word in x.split() if word not in stop_words]))
    data = data[data['word_count'] > 0]

    for article_data in data.to_dict('records'):
        db.articles_cleaned.update_one(
            {'title': article_data['title'], 'link': article_data['link']},
            {'$set': article_data},
            upsert=True
        )

    return data

if __name__ == '__main__':
    clean_data()


reader/app.py
from flask import Flask, render_template, request, jsonify
from pymongo import MongoClient
from joblib import load
import feedparser
from urllib.parse import urlparse
from datetime import datetime
from dateutil import parser
import pytz
from bs4 import BeautifulSoup
from data_cleaner import clean_text
import random
import requests
from flask_caching import Cache

app = Flask(__name__)

cache = Cache(config={'CACHE_TYPE': 'simple'})
cache.init_app(app)

client = MongoClient('mongodb://localhost:27017/')
db = client['cybersecurity_news']

model = load('model.pkl')

def get_credibility_score(source):
    credibility_scores = {
        'The Hacker News': 9,
        'Graham Cluley': 9.5,
        'Naked Security': 8,
        'Infosecurity Magazine': 8.5,
        'Cyberscoop': 8,
        'SecurityWeek': 9,
        'CSO Online': 8.5,
        'ZDNet Zero Day': 8,
        'Schneier on Security': 9.5
    }
    return credibility_scores.get(source, 5)

def get_freshness_score(published_time):
    current_time = datetime.now(pytz.utc)
    published_time = datetime(*published_time[:6], tzinfo=pytz.utc)
    time_difference = current_time - published_time
    hours_difference = time_difference.total_seconds() / 3600
    return max(10 - hours_difference / 24, 0)

def get_engagement_score():
    return 5

def get_overall_score(credibility, freshness, engagement):
    return credibility * 0.5 + freshness * 0.3 + engagement * 0.2

def extract_thumbnail(entry):
    thumbnail = None

    # Check if the entry has a 'media_thumbnail' field
    if 'media_thumbnail' in entry and len(entry.media_thumbnail) > 0:
        thumbnail = entry.media_thumbnail[0]['url']

    # Check if the entry has 'enclosures'
    elif 'links' in entry:
        for link in entry.links:
            if link.rel == 'enclosure' and link.type.startswith('image'):
                thumbnail = link.href
                break

    # If no thumbnail found, retrieve a random image from Pixabay
    if not thumbnail:
        api_key = "your_api_key"
        query = ["technology", "computing", "cybersecurity", "encryption", "firewall", "malware", "phishing", 
                 "ransomware", "spyware", "threat", "vulnerability", "intrusion", "forensics", "network", 
                 "data", "privacy", "protection", "breach", "alert", "defense", "identity", "risk", "scam", 
                 "security", "attack", "cybercrime", "hacker", 
                 "safe", "secure", "system", "information", "internet", "virus", "web", "password"]
        thumbnail = get_random_image_from_pixabay(api_key, query)

    return thumbnail

# Create a set to store used image URLs
used_images = set()

def get_random_image_from_pixabay(api_key, query):
    url = "https://pixabay.com/api/"
    params = {
        "key": api_key,
        "q": query,
        "orientation": "horizontal",
        "per_page": 20
    }
    response = requests.get(url, params=params)
    if response.status_code == 200:
        result = response.json()
        if "hits" in result and len(result["hits"]) > 0:
            # Select a random image and check if it's been used
            image = random.choice(result["hits"])
            while image["largeImageURL"] in used_images and len(used_images) < len(result["hits"]):
                image = random.choice(result["hits"])
            used_images.add(image["largeImageURL"])
            return image["largeImageURL"]
    return None


@cache.memoize(50)
def get_articles(offset=0, per_page=18):
    articles = []
    sources = {
        'The Hacker News': 'https://thehackernews.com/feeds/posts/default',
        'Graham Cluley': 'https://grahamcluley.com/feed/',
        'Naked Security': 'https://nakedsecurity.sophos.com/feed/',
        'The Last Watchdog': 'https://www.lastwatchdog.com/feed/',
        'The State of Security – Tripwire': 'https://www.tripwire.com/state-of-security/feed/',
        'CSO Online': 'https://www.csoonline.com/index.rss',
        'Infosecurity Magazine': 'https://www.infosecurity-magazine.com/rss/news/',
        'Cyber Magazine': 'https://cybermagazine.com/rss',
        'Security Weekly': 'https://securityweekly.com/feed/',
        'IT Security Guru': 'https://www.itsecurityguru.org/feed/'
    }

    for source_name, source_url in sources.items():
        feed = feedparser.parse(source_url)
        for entry in feed.entries:
            credibility = get_credibility_score(source_name)
            freshness = get_freshness_score(entry.published_parsed)
            engagement = get_engagement_score()
            overall_score = get_overall_score(credibility, freshness, engagement)

            title = entry.get('title', '')
            summary = entry.get('summary', '')

            cleaned_title = clean_text(title)

            if summary.startswith("http"):
                response = requests.get(summary)
                if response.status_code == 200:
                    summary = response.text
                else:
                    summary = ""

            if summary:
                if '<' in summary and '>' in summary:
                    soup = BeautifulSoup(summary, 'html.parser')
                    cleaned_summary = soup.get_text()
                else:
                    cleaned_summary = summary
            else:
                cleaned_summary = ""

            published_date = parser.parse(entry.published)

            thumbnail = extract_thumbnail(entry)

            article = {
                'source': source_name,
                'title': title,
                'cleaned_title': cleaned_title,
                'link': entry.link,
                'published': published_date,
                'summary': cleaned_summary,
                'credibility_score': credibility,
                'freshness_score': freshness,
                'engagement_score': engagement,
                'overall_score': overall_score,
                'thumbnail': thumbnail
            }

            articles.append(article)

    articles.sort(key=lambda x: x['published'], reverse=True)

    formatted_articles = [
        {
            'source': article['source'],
            'title': article['title'],
            'cleaned_title': article['cleaned_title'],
            'link': article['link'],
            'published': article['published'].strftime('%d %B'),
            'summary': article['summary'],
            'credibility_score': article['credibility_score'],
            'freshness_score': article['freshness_score'],
            'engagement_score': article['engagement_score'],
            'overall_score': article['overall_score'],
            'thumbnail': article['thumbnail']
        }
        for article in articles
    ]

    return formatted_articles[offset:offset + per_page]




@app.route('/', methods=['GET'])
def home():
    page = request.args.get('page', 1, type=int)
    per_page = 18
    offset = (page - 1) * per_page
    articles = get_articles(offset, per_page)
    total_articles = db.articles_cleaned.count_documents({})
    total_pages = total_articles // per_page
    if total_articles % per_page > 0:
        total_pages += 1
    return render_template('index.html', articles=articles, total_pages=total_pages, page=page)

@app.route('/clear_cache')
def clear_cache():
    cache.clear()
    return "Cache cleared"

@app.route('/articles/<int:page>', methods=['GET'])
def articles_page(page):
    per_page = 18
    offset = (page - 1) * per_page
    articles = get_articles(offset, per_page)
    total_articles = db.articles_cleaned.count_documents({})
    total_pages = total_articles // per_page
    if total_articles % per_page > 0:
        total_pages += 1
    return render_template('index.html', articles=articles, total_pages=total_pages, page=page)

@app.route('/predict', methods=['POST'])
def predict():
    data = request.get_json()
    article = data['article']
    prediction = model.predict([article])[0]
    return jsonify({'prediction': prediction})

@app.template_filter('format_date')
def format_date(value):
    date = datetime.strptime(value, "%Y-%m-%dT%H:%M:%S")
    return date.strftime('%d %B')

@app.template_filter('truncate_words')
def truncate_words(s, num=20):
    words = s.split()
    if len(words) > num:
        words = words[:num]
    return ' '.join(words)

if __name__ == '__main__':
    app.run()
